### PR TITLE
Issue 254 | Gateway service

### LINKED ISSUE #254: Users: A way to limit number of jobs per user
### What is the expected behavior?

When we have keyclock integrated, we need to define a way to configure the maximum number of jobs allowed to run in our ray pods by user, to have a mechanism to control the saturation of the system.

We can have a number of jobs allowed to run at the same time by role, so depending of the user, we can provide more flexibility.

### PR DESCRIPTION
**Pull request by IceKhan13**
### Summary

Gateway service initial works.
Service for managing programs execution, setting limits, etc.

TODOs:
- [x] POC gateway
- [x] client integration
- [x] UML diagram
- [x] tests
- [x] docs
- [x] linter
- [x] update tutorials
- [x] commands to create user, compute resource, site, etc.
- [x] dockerfile
- [x] CI tests
- [x] job results

### Details and comments

```python
from quantum_serverless import QuantumServerless, Program, GatewayProvider, Provider
from quantum_serverless.core import ComputeResource

gateway_provider = GatewayProvider(
    username="john",
    password="password123",
    # token="<TOKEN>"  # alternatevely can provide token instead of user/password
    host="http://localhost:8000",
)

serverless = QuantumServerless([gateway_provider])

program = Program(
    name="awesome_program",
    entrypoint="program.py",
    working_dir="./",
    description="description of a program",
    version="0.0.1"
)

job = serverless.run_program(program)

job.logs()
job.result()
```

Requests flow

all_diagrams-serverless_ UML drawio (1)

General overview

all_diagrams-Gateway overview drawio

Closes #254

### CODE REVIEW

**Comment by IceKhan13** [comment]
Number of things is still need to be polished and finished, but it will be separate PRs. Majority of work is done here.

Remaining things:
- CI for nightly builds
- more tests coverage
- improve code quality (serialization, verification django)
- helm chart

File: gateway/api/views.py [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+    @action(methods=["POST"], detail=False)
+    def run_program(self, request):
+        """Runs provided program on compute resources."""
+
+        serializer = self.get_serializer(data=request.data)
+        if serializer.is_valid():
+            # create program
+            program = Program(**serializer.data)
+            _, dependencies = try_json_loads(program.dependencies)
+            _, arguments = try_json_loads(program.arguments)
```
Comment (Tansito) [inline gateway/api/views.py:51]: Can this be solved with the [validators](https://github.com/Qiskit-Extensions/quantum-serverless/tree/main/repository/api/validators)?

File: gateway/gateway/settings.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"
+
+REST_FRAMEWORK = {
+    # Use Django's standard `django.contrib.auth` permissions,
+    # or allow read-only access for unauthenticated users.
+    "DEFAULT_PERMISSION_CLASSES": [
+        "rest_framework.permissions.DjangoModelPermissionsOrAnonReadOnly"
+    ],
+    "DEFAULT_AUTHENTICATION_CLASSES": (
+        "rest_framework_simplejwt.authentication.JWTAuthentication",
```
Comment (Tansito) [inline gateway/gateway/settings.py:150]: I would need to read more about `dj_rest` but probably default configuration from django can be removed if we are not going to use it.

File: gateway/gateway/urls.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+router.register(r"jobs", JobViewSet)
+
+
+urlpatterns = [
+    path("", include(router.urls)),
+    path("dj-rest-auth/", include("dj_rest_auth.urls")),
+    path("dj-rest-auth/keycloak/", KeycloakLogin.as_view(), name="keycloak_login"),
+    path("dj-rest-auth/keycloak/login/", KeycloakUsersView.as_view()),
+    path("accounts/", include("allauth.urls")),
+    path("api-auth/", include("rest_framework.urls")),
```
Comment (Tansito) [inline gateway/gateway/urls.py:33]: Similar to a previous comment. Default end-points from django can be removed but I will analyze it.

File: gateway/gateway/__init__.py
Comment (Tansito) [inline gateway/gateway/__init__.py:1]: In the `repository` we called this folder `main`. I think it has more sense.

**Review by Tansito — COMMENTED** [review]
This job is awesome @IceKhan13 . I couldn't review correctly all the code but I started with the `django` part. I left some comments with doubts and another ones to discuss. Something in what we can work too is to sync a little bit the projects from `repository` and this one.

File: gateway/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+COPY ./requirements.txt .
+RUN pip install -r requirements.txt
+
+# copy entrypoint.sh
+COPY ./entrypoint.sh .
+RUN sed -i 's/\r$//g' /usr/src/app/entrypoint.sh
+RUN chmod +x /usr/src/app/entrypoint.sh
+
+# copy project
+COPY . .
```
Comment (akihikokuroda) [inline gateway/Dockerfile:27]: This COPY overwrites the edited `entrypoint.sh`, doesn't it?

Comment (IceKhan13) [reply gateway/Dockerfile:27]: good point!

File: gateway/api/migrations/0001_initial.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
+# Generated by Django 4.1 on 2023-03-13 14:52
```
[the remark this answers is not in the document]
Comment (IceKhan13) [reply gateway/api/migrations/0001_initial.py:1]: That is a good point! We will do that. Moreover later on we will need to migrate to postgres as backend which will require another set of migrations.

File: gateway/api/models.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
+"""Models."""
+from django.core.validators import FileExtensionValidator
+from django.db import models
+from django.conf import settings
+
+
+class Program(models.Model):
+    """Program model."""
+
+    title = models.CharField(max_length=255)
```
[the remark this answers is not in the document]
Comment (IceKhan13) [reply gateway/api/models.py:10]: yes, good suggestions! and we also need UID like in repository

File: gateway/api/serializers.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+from api.models import Program, Job
+
+
+class ProgramSerializer(serializers.ModelSerializer):
+    """ProgramSerializer."""
+
+    class Meta:
+        model = Program
+        fields = ["title", "entrypoint", "artifact", "dependencies", "arguments"]
```
[the remark this answers is not in the document]
Comment (IceKhan13) [reply gateway/api/serializers.py:13]: Let's add validation as a follow up PR

File: gateway/api/views.py (hunk 1 above)
Comment (IceKhan13) [reply gateway/api/views.py:51]: Let's add validation as a follow up PR

**Comment by psschwei** [comment]
Whoops, I should've looked closer at the PR and not tried using the helm install here... will pop back over to docker-compose to poke around.

File: gateway/gateway/settings.py (hunk 2 above)
Comment (IceKhan13) [reply gateway/gateway/settings.py:150]: Let's get back to it later.

File: gateway/gateway/__init__.py
Comment (IceKhan13) [reply gateway/gateway/__init__.py:1]: good suggestion!

**Comment by IceKhan13** [comment]
@psschwei btw, docker compose will be updated with https://github.com/Qiskit-Extensions/quantum-serverless/pull/301 to latest Keycloak image

**Comment by pacomf** [comment]
@psschwei the idea for the near future is use the gateway like entry point (to connect with repository, keycloak, etc), to we will have just one interface (API Gateway) to connect from Middleware client

**Comment by akihikokuroda** [comment]
When is the Job in the gateway deleted?  I haven't looked though changes yet so I might miss it.

**Comment by IceKhan13** [comment]
it remains there for references, until user decides to delete it @akihikokuroda . We can put ttl or something like this later on

**Review by Tansito — APPROVED** [review]
As we talked this is a good first approach. From my part is approved although it will require some improvements that we will cover in future issues / pull requests:

- #307 
- #308
- #310 
- #311
- #312
- #314
- #315
- Add in `repository` validation for artifactory

I wil create them tomorrow to start handle them.

**Review by pacomf — APPROVED** [review]
LGTM!

File: infrastructure/helm/quantumserverless/charts/gateway/Chart.yaml [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+# This is the chart version. This version number should be incremented each time you make changes
+# to the chart and its templates, including the app version.
+# Versions are expected to follow Semantic Versioning (https://semver.org/)
+version: 0.1.0
+
+# This is the version number of the application being deployed. This version number should be
+# incremented each time you make changes to the application. Versions are not expected to
+# follow Semantic Versioning. They should reflect the version the application is using.
+# It is recommended to use it with quotes.
+appVersion: "1.16.0"
```
Comment (akihikokuroda) [inline infrastructure/helm/quantumserverless/charts/gateway/Chart.yaml:24]: Is "0.1.0" better?

File: infrastructure/helm/quantumserverless/values.yaml [hunk 9]
```diff
@@ -9,6 +9,28 @@
...
+gatewayEnable: true
+gateway:
+  image:
+    pullPolicy: Always
+  application:
+    siteHost: "http://127.0.0.1:8000"
+    rayHost: "http://serverless-cluster-kuberay-head-svc:8265"
+    keycloak:
+      clientId: "gateway-client"
+      url: "http://serverless-cluster-keycloak/"
```
Comment (akihikokuroda) [inline infrastructure/helm/quantumserverless/values.yaml:25]: Is `serverless-cluster` your helm install release name?  Would you replace it with HELM-RELEASE

File: infrastructure/helm/quantumserverless/values.yaml [hunk 10]
```diff
@@ -9,6 +9,28 @@
...
+      realm: "quantumserverless"
+      clientSecret: "secret"
+      clientName: "gateway-client"
+    superuser:
```
Comment (akihikokuroda) [inline infrastructure/helm/quantumserverless/values.yaml:29]: Is this used?

**Review by akihikokuroda — APPROVED** [review]
LGTM.  Great work!  There are some minor comments but I approve this.

**Pull request merged**